engines/native/system/process_posix.go

```diff
@@ -164,36 +155,59 @@ func StartProcess(options ProcessOptions) (*Process, error) {
 	return p, nil
 }
 
-// KillByOwner will kill all process with the given owner.
-func KillByOwner(user *User) error {
-	// Create pkill command
-	cmd := exec.Command(systemPKill, "-u", user.name)
-	var out bytes.Buffer
-	cmd.Stdout = &out
-	cmd.Stderr = &out
+func (p *Process) waitForResult() {
+	err := p.cmd.Wait()
+	debug("Process, cmd.Wait() return: %v", err)
 
-	// Start pkill, error here is pretty fatal
-	err := cmd.Start()
-	if err != nil {
-		panic(fmt.Sprintf("Failed to start pkill, error: %s", err))
+	p.sockets.Wait()
+
+	// Close all connected streams, ignore errors
+	if p.stdin != nil {
+		p.stdin.Close()
+	}
+	if p.stdout != nil {
+		p.stdout.Close()
+	}
+	if p.stderr != nil && p.stderr != p.stdout {
+		p.stderr.Close()
+	}
+	if p.pty != nil {
+		p.pty.Close()
 	}
 
-	// Wait for pkill to terminate
-	err = cmd.Wait()
+	// Resolve with result
+	p.resolve.Do(func() {
+		p.result = err == nil
+	})
+}
 
-	// Check error, exitcode 1 means nothing was killed we ignore that
-	if err != nil {
-		if e, ok := err.(*exec.ExitError); ok {
-			if status, ok := e.Sys().(syscall.WaitStatus); ok {
-				// Exit status 1 means no processes were killed, that's fine
-				if status.ExitStatus() == 1 {
-					return nil
-				}
-			}
-			return fmt.Errorf("pkill exited non-zero with output: %s", out.String())
-		}
-		return fmt.Errorf("pkill error: %s", err)
+// Wait for process to terminate, returns true, if exited zero.
+func (p *Process) Wait() bool {
+	p.resolve.Wait()
+	return p.result
+}
+
+// Kill the process
+func (p *Process) Kill() {
+	p.cmd.Process.Kill()
+}
+
+// SetSize of the TTY, if running as TTY or do nothing.
+func (p *Process) SetSize(columns, rows uint16) {
+	if p.pty != nil {
+		p.pty.SetSize(columns, rows)
 	}
+}
 
-	return nil
+// KillByOwner will kill all process with the given owner.
+func KillByOwner(user *User) error {
+	// Create pkill command
+	uid := strconv.FormatUint(uint64(user.uid), 10)
+	return pkill("-u", uid)
+}
+
+// KillChildren will kill all process that are child of the given process.
+func KillChildren(process *Process) error {
+	pid := strconv.Itoa(process.cmd.Process.Pid)
+	return pkill("-P", pid)
```

Comment:
from man:
> -P ppid,...
> Only match processes whose parent process ID is listed.

So what about grand children ?

We might have to use process groups see:
https://medium.com/@felixge/killing-a-child-process-and-all-of-its-children-in-go-54079af94773#.dft0n4rey

But I suspect the safe thing to do is: `SIGSTOP` then traverse children... and `SIGSTOP` all of them recursively... Until you hit the leafs of the trees and then do `SIGKILL` for all of them.